discovery.go

```diff
@@ -103,3 +117,48 @@ func exists(path string) bool {
 	_, err := os.Stat(path)
 	return err == nil || !os.IsNotExist(err)
 }
+
+type commandDescription struct {
+	Name     string              `json:"name"`
+	Summary  string              `json:"summary"`
+	Commands commandDescriptions `json:"commands,omitempty"`
+}
+
+type commandDescriptions []*commandDescription
+
+func (d *commandDescription) ToCommand(parent Module, discoveredIn, executable string, args []string, depth, maxDepth int) Command {
+	c := &executableCommand{
+		parent:       parent,
+		discoveredIn: discoveredIn,
+		path:         executable,
+		args:         args,
+		name:         d.Name,
+		summary:      d.Summary,
+	}
+
+	if len(d.Commands) > 0 && (maxDepth == -1 || depth < maxDepth) {
+		m := &executableModule{executableCommand: *c}
+		for _, child := range d.Commands {
+			m.cmds = append(m.cmds, child.ToCommand(m, discoveredIn, executable, append(args, child.Name), depth+1, maxDepth))
```

Comment:
It's a little harder to read / reason about a function signature with a ton of arguments. 

It is more readable and functional to pass around the whole type / object as you had in other places.

Below is just a suggestion of what it could look like alternatively. Also attached tests which seemed to be fine afterwards as well.

```diff
diff --git a/discovery.go b/discovery.go
index dcca683..4f9eeed 100644
--- a/discovery.go
+++ b/discovery.go
@@ -81,15 +83,29 @@ func (d *discoverer) discoverIn(p string, parent Module, all *Commands) {
                                if tree, err := inspect(path.Join(p, name)); err != nil {
                                        d.onError(DiscoveryError{Cause: err, Path: path.Join(p, name)})
                                } else {
-                                       *all = append(*all, tree.ToCommand(parent, p, path.Join(p, name), nil, d.depth, d.maxDepth))
+                                       *all = append(
+                                               *all,
+                                               tree.ToCommand(
+                                                       &executableCommand{
+                                                               parent:       parent,
+                                                               discoveredIn: p,
+                                                               path:         path.Join(p, name),
+                                                       },
+                                                       d.depth,
+                                                       d.maxDepth,
+                                               ),
+                                       )
@@ -126,25 +142,35 @@ type commandDescription struct {
 
 type commandDescriptions []*commandDescription
 
-func (d *commandDescription) ToCommand(parent Module, discoveredIn, executable string, args []string, depth, maxDepth int) Command {
-       c := &executableCommand{
-               parent:       parent,
-               discoveredIn: discoveredIn,
-               path:         executable,
-               args:         args,
-               name:         d.Name,
-               summary:      d.Summary,
-       }
+func (d *commandDescription) ToCommand(
+       c *executableCommand,
+       depth int,
+       maxDepth int,
+) Command {
+       c.name = d.Name
+       c.summary = d.Summary
 
        if len(d.Commands) > 0 && (maxDepth == -1 || depth < maxDepth) {
                m := &executableModule{executableCommand: *c}
                for _, child := range d.Commands {
-                       m.cmds = append(m.cmds, child.ToCommand(m, discoveredIn, executable, append(args, child.Name), depth+1, maxDepth))
+                       cc := &executableCommand{
+                               parent:       m,
+                               path:         c.path,
+                               name:         child.Name,
+                               summary:      child.Summary,
+                               args:         append(c.args, child.Name),
+                               discoveredIn: c.discoveredIn,
+                       }
+
+                       m.cmds = append(
+                               m.cmds,
+                               child.ToCommand(cc, depth+1, maxDepth),
+                       )
                }
                return m
-       } else {
-               return c
        }
+
+       return c
 }
```



```
go test ./... -v
=== RUN   TestFind
--- PASS: TestFind (0.00s)
=== RUN   TestFlatten
--- PASS: TestFlatten (0.00s)
=== RUN   TestCompletionsFor
--- PASS: TestCompletionsFor (0.00s)
=== RUN   TestDiscoverIn
--- PASS: TestDiscoverIn (0.03s)
=== RUN   TestFollowingSymlinks
--- PASS: TestFollowingSymlinks (0.01s)
=== RUN   TestGetMessageFromExecutionHandlesErrors
--- PASS: TestGetMessageFromExecutionHandlesErrors (0.01s)
=== RUN   TestGetMessageFromExecutionTrimsLineBreaks
--- PASS: TestGetMessageFromExecutionTrimsLineBreaks (0.01s)
=== RUN   TestGetMessageFromExecutionWithoutArgs
--- PASS: TestGetMessageFromExecutionWithoutArgs (0.01s)
=== RUN   TestGetMessageFromExecutionWithArgs
--- PASS: TestGetMessageFromExecutionWithArgs (0.01s)
=== RUN   TestNormalizeArgs
--- PASS: TestNormalizeArgs (0.00s)
=== RUN   TestIdentify
: no such command x
Did you mean?
   a
   b
: no such command b x
Did you mean?
   b
   b c
   b d
: no such command b d x
Did you mean?
   b d
   b d e
--- PASS: TestIdentify (0.00s)
=== RUN   TestBuildMenuUsage
--- PASS: TestBuildMenuUsage (0.00s)
=== RUN   TestBuildMenuTrailer
--- PASS: TestBuildMenuTrailer (0.00s)
=== RUN   TestBuildMenuSectionsUncached
--- PASS: TestBuildMenuSectionsUncached (0.02s)
=== RUN   TestBuildMenuSectionsReadFromCache
--- PASS: TestBuildMenuSectionsReadFromCache (0.00s)
=== RUN   TestBuildMenuSectionsWriteToCacheWhenStale
--- PASS: TestBuildMenuSectionsWriteToCacheWhenStale (0.00s)
=== RUN   TestBuildMenuSectionsWriteToCacheWhenMissing
--- PASS: TestBuildMenuSectionsWriteToCacheWhenMissing (0.00s)
=== RUN   TestIsNull
--- PASS: TestIsNull (0.00s)
=== RUN   TestSuggestionsFor
--- PASS: TestSuggestionsFor (0.00s)
=== RUN   TestUsageRelativeTo
--- PASS: TestUsageRelativeTo (0.00s)
PASS
ok      github.com/square/exoskeleton   (cached)
=== RUN   TestMarshal
--- PASS: TestMarshal (0.00s)
PASS
ok      github.com/square/exoskeleton/pkg/shellcomp     (cached)

```

Reply:
Good call. I applied this idea — passing `*executableCommand` in — and iterated a bit more.

I made this a function on `*discoverer` so that `maxDepth` (which never changed) didn't have to be passed to everything as well.

And passing `depth` was redundant with passing `*executableCommand` and `args`